src/hiveboard_bridge/include/hiveboard_bridge/TCPServer.h

```diff
@@ -0,0 +1,66 @@
+#ifndef CATKIN_ROS_TCPSERVER_H
+#define CATKIN_ROS_TCPSERVER_H
+
+#include "ITCPServerMonitor.h"
+#include "ros/ros.h"
+#include <cstring>
+#include <netinet/in.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#define TCP_BUFFER_LENGTH 65536 // max value of uint16
+
+class TCPServer {
+  public:
+    /**
+     * Construct a TCPServer
+     * @param port Port to open the server to
+     * @param monitor A monitor object to orchestrate the socket
+     */
+    TCPServer(int port, ITCPServerMonitor& monitor);
+    ~TCPServer();
+
+    /**
+     * Start listening for incoming client connections and accept one client. This is blocking until
+     * a client connects.
+     */
+    void listen();
+
+    /**
+     * Read data from socket.
+     * @param buff The buffer on which the incoming data will be written.
+     * @param length The length of the provided buffer.
+     * @param blocking Whether the call should block until some data is available.
+     * @return The length of the data read.
+     */
+    int read(char* buff, const uint16_t length, bool blocking);
+
+    /**
+     * Send some data to the client.
+     * @param buff The buffer containing the data to send.
+     * @param length The length of the data on the buffer.
+     */
+    void send(char* buff, const uint16_t length);
+
+    /**
+     * Terminate a client connection.
+     */
+    void close();
+
+  private:
+    int m_serverFd, m_clientFd, m_port;
+    int m_addressLength;
+    struct sockaddr_in m_address;
+    char m_buffer[TCP_BUFFER_LENGTH];
+
+    ITCPServerMonitor& m_monitor;
```

Comment:
Should we use a smart pointer here instead of a reference? We use reference in the hivemind since everything is on the stack, but here, idk if a shared_pointer would be more appropriate. What is the lifetime of the server monitor?

You could probably throw on a NULL pointer if you want, or allow a null pointer and not use the monitor

Reply:
Another question, does the monitor really need to be injected here? Could you manage this logic highet by changing init and listen to return a bool on success, then add the socket that succeeded to the monitor?

Reply:
The lifetime of the monitor is about the same as the lifetime of the `TCPServer`. In fact, in the first versions of my code, the calls to the `FD_xxx` macros and `select()` was directly in the body of the server's methods. I simply separated this into two classes to separate the concerns.

Actually, using injection here is a bad design I think since there is a composition relationship between the server and the monitor (a server _has a_ monitor). I'll change that so that the server creates the monitor by itself.

Reply:
Wouldn't a monitor have a server? I though a monitor could manage multiple server, if you do that it would be a one to one relationship? And again, why wouldn't a monitor contain a server instead of a server containing a monitor? 

The monitor should manage the monitoree (or the monitoree be injected if not managed) right? I'm just trying to understand the relationship.

Reply:
Perhaps the name `monitor` is confusing. Each `server` has its own `monitor`. The monitor is simply a watch list of the connected clients for a given `server` (is also comprises the server's FD). 

This watch list is used to monitor the activity on the various client sockets. In our case, we only have one client by `monitor`, so we only have one client in the watch list.

Reply:
Thanks that clarifies it!